xkafka/consumer.go

```diff
@@ -111,23 +123,100 @@ func (c *Consumer) Start(ctx context.Context) error {
 				}
 			}
 
+			msg := newMessage(c.name, km)
+
+			err = c.handler.Handle(ctx, msg)
+			if ferr := c.config.errorHandler(err); ferr != nil {
+				errChan <- ferr
+
+				continue
+			}
+
+			err = c.storeMessage(msg)
+			if err != nil {
+				errChan <- err
+
+				continue
+			}
+		}
+	}
+}
+
+func (c *Consumer) runAsync(ctx context.Context) error {
+	st := stream.New().WithMaxGoroutines(c.config.concurrency)
+	ctx, cancel := context.WithCancelCause(ctx)
+
+	for {
+		select {
+		case <-ctx.Done():
+			st.Wait()
+
+			uerr := c.unsubscribe()
+
+			err := context.Cause(ctx)
+			if err != nil && err == context.Canceled {
+				return uerr
+			}
+
+			return errors.Join(err, uerr)
+		default:
+			km, err := c.kafka.ReadMessage(c.config.pollTimeout)
+			if err != nil {
+				if kerr, ok := err.(kafka.Error); ok && kerr.Code() == kafka.ErrTimedOut {
+					continue
+				}
+
+				if ferr := c.config.errorHandler(err); ferr != nil {
+					cancel(ferr)
+
+					continue
+				}
+			}
+
+			msg := newMessage(c.name, km)
+
 			st.Go(func() stream.Callback {
-				c.runHandler(ctx, c.handler, km, errChan)
+				err := c.handler.Handle(ctx, msg)
+				if ferr := c.config.errorHandler(err); ferr != nil {
+					cancel(ferr)
 
-				return func() {}
+					return func() {}
+				}
+
+				return func() {
+					if err := c.storeMessage(msg); err != nil {
+						cancel(err)
+					}
+				}
 			})
 		}
 	}
 }
 
-func (c *Consumer) runHandler(ctx context.Context, handler Handler, km *kafka.Message, errChan chan error) {
-	msg := newMessage(c.name, km)
+func (c *Consumer) storeMessage(msg *Message) error {
+	if msg.Status != Success && msg.Status != Skip {
+		return nil
+	}
 
-	err := handler.Handle(ctx, msg)
+	_, err := c.kafka.StoreOffsets([]kafka.TopicPartition{
```

Comment:
Offset storage is quite unintuitive to reason about, can we add some documentation with few examples around this?
Explaining its behaviour with both sync/async mode of consumption and what does calling `storeMessage` async mean.

Some pieces are not trivial to make sense of, for example, why is it `Offset: kafka.Offset(msg.Offset + 1)` and not `Offset: kafka.Offset(msg.Offset)`?

Reply:
> why is it Offset: kafka.Offset(msg.Offset + 1) and not Offset: kafka.Offset(msg.Offset)?

This is based on https://github.com/confluentinc/confluent-kafka-go/blob/af4a5f8b2018db6503f7e8097a25a24a6d6feb06/kafka/consumer.go#L374

it ensures that the consumer starts with next message when it restarts

Reply:
Okay, let's add this as a doc comment at relevant line in code for other readers.